tracing/src/main/java/com/palantir/tracing/DeferredTracer.java

```diff
@@ -49,26 +66,48 @@ public DeferredTracer(String operation) {
         this(Optional.of(operation));
     }
 
-    DeferredTracer(Optional<String> operation) {
-        this.trace = Tracer.copyTrace();
-        this.operation = operation;
+    /**
+     * Create a new deferred tracer, optionally specifying an operation.
+     * If no operation is specified, will attempt to use the parent span's operation name.
+     */
+    public DeferredTracer(Optional<String> operation) {
+        Optional<Trace> maybeTrace = Tracer.copyTrace();
+        if (maybeTrace.isPresent()) {
+            Trace trace = maybeTrace.get();
+            this.traceId = trace.getTraceId();
+            this.isObservable = trace.isObservable();
+            this.parentSpanId = trace.top().map(OpenSpan::getSpanId).orElse(null);
+            this.operation = operation.orElse(trace.top().map(OpenSpan::getOperation).orElse(DEFAULT_OPERATION));
+        } else {
+            this.traceId = null;
+            this.isObservable = false;
+            this.parentSpanId = null;
+            this.operation = null;
+        }
     }
 
     /**
      * Runs the given callable with the current trace at
      * the time of construction of this {@link DeferredTracer}.
      */
     public <T, E extends Throwable> T withTrace(Tracers.ThrowingCallable<T, E> inner) throws E {
-        if (!trace.isPresent()) {
+        if (traceId == null) {
             return inner.call();
         }
+
         Optional<Trace> originalTrace = Tracer.copyTrace();
-        Tracer.setTrace(trace.get());
-        operation.ifPresent(Tracer::startSpan);
+
+        Tracer.setTrace(new Trace(isObservable, traceId));
```

Comment:
By creating a new trace here we lose the callers span stack. While the span tree derived from trace logs is correct because we set the `parentSpanId`, the span stack available at runtime is missing parent spans.

Reply:
yes, this was a known change - I'm curious what you're doing with the span stack at runtime

Reply:
I think this breaks an assumption that the current span stack is present in the trace. It seems dangerous and confusing for the span stack to not have the full trace information.

For example, after this PR it can be valid to have a trace with an empty span stack since this trace might actually have spans that have been dropped by the deferred tracer.